sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,774 @@
+        neighborhood_size = min_neighborhood_size
+
+    local_maxima_points = _find_local_maxima(reachability_plot,
+                                             reachability_ordering,
+                                             neighborhood_size)
+    root_node = _TreeNode(reachability_ordering, 0,
+                          len(reachability_ordering), None)
+    _cluster_tree(root_node, None, local_maxima_points,
+                  reachability_plot, reachability_ordering, min_cluster_size,
+                  maxima_ratio, rejection_ratio,
+                  similarity_threshold, significant_min)
+
+    return root_node
+
+
+class _TreeNode(object):
+    # automatic cluster helper classes and functions
+    def __init__(self, points, start, end, parent_node):
+        self.points = points
+        self.start = start
+        self.end = end
+        self.parent_node = parent_node
+        self.children = []
+        self.split_point = -1
+
+    def assign_split_point(self, split_point):
+        self.split_point = split_point
+
+    def add_child(self, child):
+        self.children.append(child)
+
+
+def _is_local_maxima(index, reachability_plot, reachability_ordering,
+                     neighborhood_size):
+    # 0 = point at index is not local maxima
+    # 1 = point at index is local maxima
+    for i in range(1, neighborhood_size + 1):
+        # process objects to the right of index
+        if index + i < len(reachability_plot):
+            if (reachability_plot[index] < reachability_plot[index + i]):
+                return 0
+        # process objects to the left of index
+        if index - i >= 0:
+            if (reachability_plot[index] < reachability_plot[index - i]):
+                return 0
+    return 1
+
+
+def _find_local_maxima(reachability_plot, reachability_ordering,
+                       neighborhood_size):
+    local_maxima_points = {}
+    # 1st and last points on Reachability Plot are not taken
+    # as local maxima points
+    for i in range(1, len(reachability_ordering) - 1):
+        # if the point is a local maxima on the reachability plot with
+        # regard to neighborhood_size, insert it into priority queue and
+        # maxima list
+        if (reachability_plot[i] > reachability_plot[i - 1] and
+            reachability_plot[i] >= reachability_plot[i + 1] and
+            _is_local_maxima(i, reachability_plot, reachability_ordering,
+                             neighborhood_size) == 1):
+            local_maxima_points[i] = reachability_plot[i]
+
+    return sorted(local_maxima_points,
+                  key=local_maxima_points.__getitem__, reverse=True)
+
+
+def _cluster_tree(node, parent_node, local_maxima_points,
+                  reachability_plot, reachability_ordering,
+                  min_cluster_size, maxima_ratio, rejection_ratio,
+                  similarity_threshold, significant_min):
+    # node is a node or the root of the tree in the first call
```

Comment:
This function could benefit from a proper docstring.

Reply:
Updated docstring... since the function is private I'm not sure if you want it to encompass all the input arguments (they're somewhat repetitively already documented earlier in the file).